src/proto/rng.rs

```diff
@@ -0,0 +1,60 @@
+//! `Rng` protocol.
+
+use crate::{data_types::Guid, proto::Protocol, unsafe_guid, Result, Status};
+use core::ptr;
+
+/// Contain a Rng algorithm Guid
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct RngAlgorithm(pub Guid);
+
+impl RngAlgorithm {
+    /// Get an empty `RngAlgorithm`
+    ///
+    /// Used provide a buffer to `Rng.get_info`
+    pub fn default() -> Self {
+        Self(Guid::default())
+    }
+}
+
+/// Rng protocol
+#[repr(C)]
+#[unsafe_guid("3152bca5-eade-433d-862e-c01cdc291f44")]
+#[derive(Protocol)]
+pub struct Rng {
+    get_info: extern "efiapi" fn(
+        this: &Rng,
+        algorithm_list_size: *mut usize,
+        algorithm_list: *mut RngAlgorithm,
+    ) -> Status,
+    get_rng: extern "efiapi" fn(
+        this: &Rng,
+        algorithm: *const RngAlgorithm,
+        value_length: usize,
+        value: *mut u8,
+    ) -> Status,
+}
+
+impl Rng {
+    /// Returns information about the random number generation implementation.
+    pub fn get_info(&mut self, algorithm_list: &mut [RngAlgorithm]) -> Result<usize> {
+        let algorithm_list_size = (algorithm_list.len() * 16) as *mut usize;
+
+        (self.get_info)(self, algorithm_list_size, algorithm_list.as_mut_ptr())
+            .into_with_val(|| algorithm_list_size as usize / 16)
```

Comment:
For UEFI functions like this one that read data into a pre-allocated buffer, and return `BUFFER_TOO_SMALL` if the buffer isn't big enough, we usually return the required size as an optional error value. Here's what it could look like:

```rust
pub fn get_info<'buf>(
    &mut self,
    algorithm_list: &'buf mut [RngAlgorithm],
) -> Result<&'buf [RngAlgorithm], Option<usize>> {
    let mut algorithm_list_size = algorithm_list.len() * mem::size_of::<RngAlgorithm>();

    (self.get_info)(self, &mut algorithm_list_size, algorithm_list.as_mut_ptr()).into_with(
        || {
            let len = algorithm_list_size / mem::size_of::<RngAlgorithm>();
            &algorithm_list[..len]
        },
        |status| {
            if status == Status::BUFFER_TOO_SMALL {
                Some(algorithm_list_size)
            } else {
                None
            }
        },
    )
}
```

The changes here are roughly:
* In the success case, the result contains a reference to the part of the buffer that's been initialized. This is just a slightly more convenient interface than returning the number of algorithms. A `'buf` lifetime has been added to tie the input and output buffer lifetimes together.
* In the error case, the result contains an optional usize representing the required size of the buffer. That option is only filled in if the status is `BUFFER_TOO_SMALL`, otherwise something else went wrong and the required buffer size isn't known.

Reply:
I never had to use `'buf` before, thanks for explain it.
Just implemented your whole function with the `unsafe` block.

Reply:
As we pass a mutable reference to the wrapper and function, we don't need to return the list. I think it should be more useful to return the variable `len` as we need to know as much algorithms has been returned.

Reply:
Either interface does work, but what I'm suggesting is to return a subslice of the original list. The user would likely end up doing that anyway right after calling the function if we just return a `usize`, so might as well take care of that for them.

Reply:
Oh right I see it.

There you go, it should be my final commit if it's good for you